common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java

```diff
@@ -64,27 +65,71 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
         SessionFactoryContext context = contextProvider.get();
 
         try {
-            SessionFactory currentSessionFactory = context.getSessionFactory();
-            if(currentSessionFactory != null) {
+            SessionFactory currentSessionFactory = context.getCurrentSessionFactory();
+            if (currentSessionFactory != null) {
                 session = currentSessionFactory.getCurrentSession();
             }
-        } catch (HibernateException e) {}
+        } catch (HibernateException e) {
+        }
 
         if (session == null) {
-            //start a new session and transaction if current session is null
-            //get DataSourceType first
-            SelectDataSource selectedDS = invocation.getMethod().getAnnotation(SelectDataSource.class);
-            if(selectedDS == null) {
-                context.useDefault();
-            } else {
-                context.useSessionFactory(selectedDS.value());
+            //start a new session if current session is null
+
+            //get shardKey from method argument if there is any
+            String shardKey;
+            ShardId shardId;
+            SessionFactory sessionFactory = null ;
+
+            try {
+                DataStorage dataStorage = invocation.getMethod().getAnnotation(DataStorage.class);
+                switch (dataStorage.value()) {
+                    case SHARDED: {
+                        try {
+                            SelectDataSource selectedDS = invocation.getMethod().getAnnotation(SelectDataSource.class);
```

Comment:
instead of having two annotations, it's good to have one, which gives both details DataSourceType and DataStorage